src/Common/ZooKeeper/ZooKeeperCommon.cpp

```diff
@@ -41,19 +43,24 @@ std::string ZooKeeperRequest::toString(bool short_format) const
 
 void ZooKeeperRequest::write(WriteBuffer & out) const
 {
-    /// Excessive copy to calculate length.
-    WriteBufferFromOwnString buf;
-    Coordination::write(xid, buf);
-    Coordination::write(getOpNum(), buf);
-    writeImpl(buf);
-    Coordination::write(buf.str(), out);
+    auto request_size = Coordination::size(xid) + Coordination::size(getOpNum()) + sizeImpl();
+
+    Coordination::write(static_cast<int32_t>(request_size), out);
```

Comment:
We write request_size , but not read it in `read` function https://github.com/ClickHouse/ClickHouse/pull/68108/files#diff-ee99436b416af0a67eba2154c1d8ade09c61f408ba67e69f842aae71dab45edaR1112

Why @_@

Reply:
because `read` function is broken and we don't use it...
we read manually `length`, `xid` and than call `readImpl`
https://github.com/ClickHouse/ClickHouse/blob/master/src/Server/KeeperTCPHandler.cpp#L584